TM-31

elizakanska opened the pull request:
The User endpoints and CRUD functionality

elizakanska committed:
Added init_02.sql for user initialization

elizakanska committed:
Added BE endpoints for Users

elizakanska committed:
Added FE user endpoints (also a quick user list)

elizakanska committed:
Fixed the same entity/dto error I had with Task before

jevgenijs-jaunslavietis commented on FE/src/app/BigImportNgZorro.ts:
```diff
@@ -0,0 +1,149 @@
+
+}
+
+// This file is used so that I don't have to import each of the components whenever I want to use them,
```
If you are not using all of the components they still will be in the findle bundle increasing the bundle size for no reason

jevgenijs-jaunslavietis commented on FE/src/app/app.component.ts:
```diff
@@ -1,11 +1,12 @@
   selector: 'app-root',
   imports: [RouterLink, RouterOutlet, SharedModule],
   templateUrl: './app.component.html',
+  standalone: true,
```
Do you still need it there? I think it is true by default in newer Angular versions.

elizakanska committed:
Made minor changes based on comment (space saving and removing standalone: true)

elizakanska committed:
Made minor changes based on comment (space saving and removing standalone: true)

elizakanska committed:
Resolving merge conflicts

wanpii commented on FE/src/app/models/user.model.ts:
```diff
@@ -0,0 +1,7 @@
+export interface User{
```
space

wanpii commented on FE/src/app/pages/users/users.component.html:
```diff
@@ -0,0 +1,109 @@
+    </tr>
+    </thead>
+    <tbody>
+    <tr *ngFor="let user of filteredUsers; trackBy: trackByUserId">
```
why not use the new syntax for for and if statements? @for @if else

wanpii commented on FE/src/app/pages/users/users.component.html:
```diff
@@ -0,0 +1,109 @@
+
+        <nz-form-item>
+          <nz-form-control
+            [nzValidateStatus]="userFormGroup.get('username')?.invalid && userFormGroup.get('username')?.touched ? 'error' : ''"
```
you could create a reusable component for this, where you can pass the form control and error message, maybe?

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -0,0 +1,171 @@
+
+  searchUsers() {
+    if (this.searchQuery.trim() === '') {
+      this.filteredUsers = [...this.users];
```
Don't you want to move filtering functionality to the backend?

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -0,0 +1,171 @@
+        lastName: user.lastName
+      });
+    } else {
+      this.userFormGroup?.reset();
```
why is userFormGroup optional?

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -0,0 +1,171 @@
+
+    const userData: User = {
+      ...this.userFormGroup.value,
+      id: this.editingUser ? this.editingUser.id : 0
```
0?

wanpii commented on FE/src/app/services/user.service.ts:
```diff
@@ -0,0 +1,54 @@
+
+  constructor(private http: HttpClient) { }
+
+  private handleError(error: HttpErrorResponse) {
```
if I'm not mistaken you already have this handleError method in the taskService, right? Instead of duplicating the code you can extract it into a separate file, let's say utils.ts or something that fits (naming) and then reuse it everywhere you need

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -0,0 +1,171 @@
+  deleteUser(id: number) {
+    if (confirm("Are you sure you want to delete this user?")) {
+      const deleteSub = this.userService.deleteUser(id).pipe(
+        switchMap(() => this.userService.getUsers())
```
I don't like that you call the BE twice here, after deleting you could just return the new list of users after deletion

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -0,0 +1,171 @@
+    NzFormItemComponent,
+    NzInputDirective,
+    NzButtonComponent
+  ]
```
change detection

wanpii commented on src/main/resources/static/init_02.sql:
```diff
@@ -0,0 +1,25 @@
+-- Create schema
+CREATE SCHEMA usermanager;
```
why are you creating a separate schema? you should have a database and under the database multiple tables, like tasks, users, etc. What's the point?

elizakanska committed:
Made changes based on mentors comments

(Thank goodness I don't actually need the second schema, that felt so weird, the task is worded weird and I got confused)

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -1,38 +1,36 @@
   editingUser: User | null = null;
   userFormGroup: FormGroup;
-  searchQuery: string = '';
+  searchQuery = '';
```
why remove type definition?

wanpii replied:
the type is still removed?

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -108,18 +89,22 @@ export class UsersComponent implements OnInit, OnDestroy {
-        this.users = data;
-        this.filteredUsers = [...this.users];
+      next: (savedUser) => {
+        if (this.editingUser) {
```
what are you trying to do here? :D

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -133,13 +118,10 @@ export class UsersComponent implements OnInit, OnDestroy {
+    if (confirm('Are you sure you want to delete this user?')) {
+      const deleteSub = this.userService.deleteUser(id).subscribe({
+        next: () => {
+          this.users = this.users.filter(user => user.id !== id);
```
This works, but to avoid this in the FE why don't you just return the new list from the BE?

elizakanska committed:
Made changes based on mentors comments

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -83,28 +84,22 @@ export class UsersComponent implements OnInit, OnDestroy {
   saveUser() {
     if (this.userFormGroup.invalid) {
-      this.userFormGroup.markAllAsTouched();
+      alert('Please fill in all required fields.');
```
think about UX, what would be more natural for a user to do, fill out some fields click create/save get an alert you didn't fill out required fields, or perhaps fill out some fields, save/create button is disabled because required fields are missing, when required fields are filled button is enabled and user can save/create or 3rd option, fill out some fields, user presses create/save but gets an error message (somewhere that makes sense) indicating not all required fields are filled and can quickly add required information and create/save

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -119,13 +114,15 @@ export class UsersComponent implements OnInit, OnDestroy {
-        next: () => {
-          this.users = this.users.filter(user => user.id !== id);
+      const deleteSub = this.userService.deleteUser(id).pipe(
+        switchMap(() => this.userService.getUsers())
```
you are still calling backend twice, once for deletion once for fetching users

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -83,28 +84,22 @@ export class UsersComponent implements OnInit, OnDestroy {
+      : this.userService.addUser(userPayload);
+
+    const saveSub = userOperation$.pipe(
+      switchMap(() => this.userService.getUsers())
```
same here, two calls when it can be reduced to one

elizakanska committed:
Made changes based on mentors comments

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -88,24 +84,19 @@ export class UsersComponent implements OnInit, OnDestroy {
+    this.subscriptions.add(
+      userOperation$.subscribe({
+        next: () => {
+          this.loadUsers();
```
now you are subscribing within a subscription again :)

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -114,30 +105,25 @@ export class UsersComponent implements OnInit, OnDestroy {
-      this.subscriptions.add(deleteSub);
+      this.subscriptions.add(
+        this.userService.deleteUser(id).subscribe({
+          next: () => this.loadUsers(),
```
same here

elizakanska committed:
Made changes based on mentors comments

I have no idea how the : string keeps vanishing every time I take my eyes off it, I fear I might be going insane

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -89,9 +87,9 @@ export class UsersComponent implements OnInit, OnDestroy {
-      userOperation$.subscribe({
-        next: () => {
-          this.loadUsers();
+      userOperation$.pipe(switchMap(() => this.userService.getUsers())).subscribe({
```
I'll give you a tip (no matter how you change the FE you will still be calling BE 2 times, once for editUser/addUser one for getUsers, the other is deleteUser + getUsers, have you thought about returning the new list of users on the FE **from the BE from the same request?**

wanpii replied:
Why did you mark it as resolved? There are no changes to what I requested

elizakanska committed:
Added behaviousubject and signal

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -20,116 +23,111 @@ import { NgClass } from '@angular/common';
+    userOperation$.pipe(
+      takeUntilDestroyed(this.destroyRef),
+      finalize(() => this.isSubmitting = false)
+    ).subscribe({
```
you're not saving changes in DB?

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -20,116 +23,111 @@ import { NgClass } from '@angular/common';
-      );
+  deleteUser(id: number): void {
+    if (confirm('Are you sure?')) {
+      this.userService.deleteUser(id).pipe(
```
Why do you refuse to send a new list of users from the BE for this deleteUser request?

wanpii commented on FE/src/app/services/user.service.ts:
```diff
@@ -9,11 +10,11 @@ import { handleError } from './utils/errorHandler';
 
   getUsers(searchQuery: string = ''): Observable<User[]> {
-    return this.http.get<User[]>(`${API_URL}/users`, { params: { searchQuery } }).pipe(
+    const params = searchQuery ? { params: { searchQuery } } : {};
```
not bad 👍

wanpii commented on src/main/java/com/kanskaeliza/taskmanager/service/impl/UserServiceImpl.java:
```diff
@@ -26,54 +26,53 @@ public UserServiceImpl(UserRepository repository, @Qualifier("userMapperImpl") U
-    }
+  public UserDTO createUser(UserDTO userDTO) {
+    validateUserDTO(userDTO);
+    User newUser = mapper.toDto(userDTO);
```
It's a little weird here, why do you map twice?

wanpii replied:
You receive a DTO already, so you should map it to the object that is saved in database and save it, yet I see you're mapping the DTO from FE to a BE dto adn then the BE dto to the final object that gets saved in DB

wanpii commented on src/main/java/com/kanskaeliza/taskmanager/service/impl/UserServiceImpl.java:
```diff
@@ -26,54 +26,53 @@ public UserServiceImpl(UserRepository repository, @Qualifier("userMapperImpl") U
+    existingUser.setLastName(userDTO.getLastName());
 
-    return mapper.toUser(userDTO);
+    return mapper.toUser(repository.save(existingUser));
```
Also I'm not sure about the implications of this synatx, where you save inside the method, please investigate this and see if it has any implications

elizakanska committed:
Modified based on mentors comments

The implications of saving within the [return mapper.toUser(repository.save(existingUser));] line were bad... found a better way

wanpii commented on FE/src/app/pages/users/users.component.ts:
```diff
@@ -89,34 +89,51 @@ export class UsersComponent {
 
   deleteUser(id: number): void {
     if (confirm('Are you sure?')) {
+      this.users.update(current => current.filter(u => u.id !== id));
```
you still haven't answered :D, why do you not return the new list of users from the backend? After any action like creating, updating or deleting user?

elizakanska replied:
Because it would make the application exponentially slower with each new user added. Also that is what making the users array a signal is for, so we wouldn't have to call on the BE twice per method, right?

wanpii replied:
How would it make the application expontentially slower, fetching users from the database is very fast and simple by querying the database? It is required, imagine there are multiple users using the system at once, one user deletes User 1, but another user wants to edit User 1, the other user will get an error each time trying to update or delete User 1, because it no longer exists. We do not call BE twice when deleting, we just add to the existing functionality of the deleteUser/editUser method in Java, after deleting or uodating the user, we query the database for the new list of users and return them, all in a single request

elizakanska committed:
Modified based on mentors comments
